Fix boot on newer qemu

lf- opened the pull request:
We were previously noncompliant with the RISC-V specification, and so
was qemu: if there are no PMP registers configured, according to the
spec, any access to memory, including instruction fetch, from S- or U-
mode, should fail (see section 3.6.1 of the RISC-V privileged spec).

However, qemu was not enforcing this properly prior to
https://github.com/qemu/qemu/commit/d102f19a208. Running a qemu newer
than this would get stuck in an infinite loop hitting the exception
vector repeatedly.

This commit adds trivial support for the PMP registers, writing a value
such that all accesses from S- and U- mode have their access managed
solely by the page tables. This restores the previous behaviour.

Further, it makes the machine mode timer vector check if it receives an
unexpected interrupt if so, jump to a specific infinite loop. This
should make any potential future occurrences of interrupt issues more
obvious.

I found this bug when my own kernel (https://github.com/lf-/mu) didn't boot, and after a "fun" session with a gdb attached to qemu, I found this issue and wondered if xv6 had the same bug. Turns out it did.

Fix #76.

lf- committed:
Add trivial support for PMPs, fixes boot on newer qemu

We were previously noncompliant with the RISC-V specification, and so
was qemu: if there are no PMP registers configured, according to the
spec, any access to memory, including instruction fetch, from S- or U-
mode, should fail (see section 3.6.1 of the RISC-V privileged spec).

However, qemu was not enforcing this properly prior to
https://github.com/qemu/qemu/commit/d102f19a208. Running a qemu newer
than this would get stuck in an infinite loop hitting the exception
vector repeatedly.

This commit adds trivial support for the PMP registers, writing a value
such that all accesses from S- and U- mode have their access managed
solely by the page tables. This restores the previous behaviour.

Further, it makes the machine mode timer vector check if it receives an
unexpected interrupt if so, jump to a specific infinite loop. This
should make any potential future occurrences of interrupt issues more
obvious.

Virtual-Machine commented on kernel/start.c:
```diff
@@ -45,10 +47,36 @@ start()
+  // range encoding in PMP address and configuration registers" in the RISC-V
+  // privileged specification
+  // we set the bits such that this matches any 56-bit physical address
+  w_pmpaddr0((~0L) >> 10);
```
While these changes work on my machine to get Qemu 6.0.0 to boot, I suspect this line is not doing what you think it is doing.

The right shift has no effect on this value because it ends up as a negative value. 
```c
-1 >> any value == -1
```

lf- replied:
You are absolutely correct. It should be `~0ULL`. I'll change it and push a fix.

lf- replied:
The code is indeed accidentally functional, because the 0s in pmpaddr0[54:63] are WARL, so it is completely legit to set them to 1. I think I'll still fix it.

Virtual-Machine replied:
Yeah, I read the spec like 10x and still couldn't be sure if -1 was a legit value or not but figured I should point it out because its easy to read the line and miss the right shift intention vs effect.

lf- committed:
Fix incorrect bit math

Thanks Virtual-Machine! It was previously accidentally shifting signed
numbers which did nothing.

```
/* original */
[cling]$ #include <stdio.h>
[cling]$ unsigned long long v = ((~0L) >> 10)
(unsigned long long) 18446744073709551615
[cling]$ printf("%llx", v)
ffffffffffffffff(int) 16

/* corrected: */
[cling]$ v = ((~0ULL) >> 10)
(unsigned long long) 18014398509481983
[cling]$ printf("%llx", v)
3fffffffffffff(int) 14
```

ceciivanov commented:
> We were previously noncompliant with the RISC-V specification, and so was qemu: if there are no PMP registers configured, according to the spec, any access to memory, including instruction fetch, from S- or U- mode, should fail (see section 3.6.1 of the RISC-V privileged spec).
> 
> However, qemu was not enforcing this properly prior to [qemu/qemu@d102f19](https://github.com/qemu/qemu/commit/d102f19a208). Running a qemu newer than this would get stuck in an infinite loop hitting the exception vector repeatedly.
> 
> This commit adds trivial support for the PMP registers, writing a value such that all accesses from S- and U- mode have their access managed solely by the page tables. This restores the previous behaviour.
> 
> Further, it makes the machine mode timer vector check if it receives an unexpected interrupt if so, jump to a specific infinite loop. This should make any potential future occurrences of interrupt issues more obvious.
> 
> I found this bug when my own kernel (https://github.com/lf-/mu) didn't boot, and after a "fun" session with a gdb attached to qemu, I found this issue and wondered if xv6 had the same bug. Turns out it did.
> 
> Fix #76.

hello can I ask how I can pull that change so  I can fix the boot bug?

lf- commented:
> hello can I ask how I can pull that change so I can fix the boot bug?

```
git fetch origin pull/62/head
git merge FETCH_HEAD
```

where you replace `origin` with whatever your remote for the upstream xv6 repository is (i.e. the one pointing to https://github.com/mit-pdos/xv6-riscv).

kaashoek commented:
Closing expired pull request.
